Command mode support (#7681)
quarkusio/quarkus: docs/src/main/asciidoc/command-mode-reference.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Command Mode Applications

include::./attributes.adoc[]

> This reference covers how to write application that run and then exit.

Reviewer: ```suggestion
This reference covers how to write applications that run and then exit.
```

== Writing Command Mode Applications

There are two different approaches that can be used to implement applications
that exit.

. Implement `QuarkusApplication` and have Quarkus run this method automatically
. Implement `QuarkusApplication` and a Java main method, and use the Java main method to launch Quarkus

In this document the `QuarkusApplication` instance is referred to as the application main,
and a class with a Java main method is the Java main.

The simplest possible command mode application might appear as follows:

[source,java]
----
import io.quarkus.runtime.QuarkusApplication;

> import io.quarkus.runtime.annotations.DefaultMain;

Reviewer: ```suggestion
import io.quarkus.runtime.annotations.QuarkusMain;
```

@QuarkusMain    // <1>
public class HelloWorldMain implements QuarkusApplication {
  @Override
  public int run(String... args) throws Exception {   // <2>
    System.out.println("Hello World");
    return 10;
 }
}
----

> <1> The `@QuarkusMain` annotation to tells Quarkus that this is the main entry point.

Reviewer: ```suggestion
<1> The `@QuarkusMain` annotation tells Quarkus that this is the main entry point.
```

<2> The `run` method is invoked once Quarkus starts, and the application stops when it finishes.

If we want to use a Java main to run the application main it would look like:

[source,java]
----
import io.quarkus.runtime.Quarkus;

> import io.quarkus.runtime.annotations.DefaultMain;

Reviewer: ```suggestion
import io.quarkus.runtime.annotations.QuarkusMain;
```

different ClassLoader to the main application, so may not behave as you would
expect.

=== Multiple Main Methods

It is possible to have multiple main methods in an application, and select between them at build time.
The `@QuarkusMain` annotation takes an optional 'name' parameter, and this can be used to select the
main to run using the `quarkus.package.main-class` build time configuration option. If you don't want
to use annotations this can also be used to specify the fully qualified name of a main class.

By default the `@QuarkusMain` with no name (i.e. the empty string) will be used, and if it is not present
and `quarkus.package.main-class` is not specified then Quarkus will automatically generate a main class
that just runs the application.

NOTE: The `name` of `@QuarkusMain` must be unique (including the default of the empty string). If you
have multiple `@QuarkusMain` annotations in your application the build will fail if the names are not
unique.

=== The command mode lifecycle

When running a command mode application the basic lifecycle is as follows:

. Start Quarkus
. Run the `QuarkusApplication` main method

> . Shutdown Quarkus and exit the JVM after the main method returns

Reviewer: ```suggestion
. Shut down Quarkus and exit the JVM after the main method returns
```
